### PR TITLE
Refactor ResultsCollector out of MetricsCalculator

### PR DESCRIPTION
**Pull request by sanmai**
This PR:

- [x] Refactors `ResultsCollector` out of `MetricsCalculator`
- [x] Covered by tests

It's a lot of changed lines, but the bottom line is that :
- We had `MetricsCalculator` doing several jobs at once, and now we have `ResultsCollector` and `MetricsCalculator` doing own parts of their jobs. 
- And there's `MutationTestingResultsCollectorSubscriber` and `MutationTestingConsoleLoggerSubscriber` doing collecting and reporting, instead of the latter doing two jobs at once.

This will let us use different strategies for result accumulation. E.g. if none of the loggers expect uncovered mutants, we can drop them, and likewise can be said about killed mutants. Most of the time they're just noise. Yet final counts will have all and every mutant counted, just like before.

I don't expect any substantial performance or memory consumption change here.

~~Blocked by #1420~~

### CODE REVIEW

**Comment by sanmai** [comment]
@theofidry this is a follow-up to #1420, that's why it basically includes all of #1420.

File: src/Container.php [hunk 1]
```diff
@@ -469,11 +475,19 @@ public static function create(): self
                     $container->getConfiguration()->noProgress()
                 );
             },
+            MutationTestingResultsCollectorSubscriberFactory::class => static function (self $container): MutationTestingResultsCollectorSubscriberFactory {
+                return new MutationTestingResultsCollectorSubscriberFactory(
+                    $container->getMetricsCalculator(),
+                    // TODO ResultsCollector can have multiple strategies on results accumulations, e.g. whenever uncovered results need saving
+                    $container->getResultsCollector()
```
Comment (sanmai) [inline src/Container.php:482]: Here's the place to inject something, that will stand between `ResultsCollectorSubscriber` and actual `ResultsCollector` doing the filtering of unneeded results.

**Comment by theofidry** [comment]
I can't look into this PR in details but I'm 👍  on the theory, it is indeed an excellent way to reduce unnecessary metrics processing depending on what loggers are used.

File: src/Event/Subscriber/MutationTestingResultsCollectorSubscriber.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+use Infection\Event\MutantProcessWasFinished;
+use Infection\Metrics\Collector;
+
+/**
+ * @internal
+ */
+final class MutationTestingResultsCollectorSubscriber implements EventSubscriber
+{
+    /** @var Collector[] */
```
Comment (maks-rafalko) [inline src/Event/Subscriber/MutationTestingResultsCollectorSubscriber.php:46]:
```suggestion
    /** @var array<int, Collector> */
```

?

File: src/Event/Subscriber/MutationTestingResultsCollectorSubscriberFactory.php [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+use Infection\Metrics\Collector;
+use Symfony\Component\Console\Output\OutputInterface;
+
+/**
+ * @internal
+ */
+final class MutationTestingResultsCollectorSubscriberFactory implements SubscriberFactory
+{
+    /** @var Collector[] */
```
Comment (maks-rafalko) [inline src/Event/Subscriber/MutationTestingResultsCollectorSubscriberFactory.php:46]:
```suggestion
    /** @var array<int, Collector> */
```

File: tests/phpunit/Logger/JsonLoggerTest.php [hunk 4]
```diff
@@ -210,12 +214,13 @@ public function metricsProvider(): iterable
...
             false,
-            $this->createUncoveredMetricsCalculator(),
+            new MetricsCalculator(2),
+            $this->createUncoveredResultsCollector(),
             [
                 'stats' => [
-                    'totalMutantsCount' => 1,
+                    'totalMutantsCount' => 0,
                     'killedCount' => 0,
-                    'notCoveredCount' => 1,
```
Comment (maks-rafalko) [inline tests/phpunit/Logger/JsonLoggerTest.php:218]: could you please explain why the stats was changed?

File: src/Event/Subscriber/MutationTestingResultsCollectorSubscriber.php (hunk 2 above)
Comment (sanmai) [reply src/Event/Subscriber/MutationTestingResultsCollectorSubscriber.php:46]: We're not using the key anywhere, so more appropriate definition would be `array<Collector>`. It is also saves us from the trouble declaring the array-key type for PHPStan up the stream. And, well, `array<Collector>` is same as `Collector[]`.

File: src/Metrics/MetricsCalculator.php [hunk 5]
```diff
@@ -44,28 +43,14 @@
...
-
-    private int $escapedCount = 0;
-
-    private int $timedOutCount = 0;
-
-    private int $notCoveredByTestsCount = 0;
+    /**
+     * @var array<string, int>
+     */
+    private array $countByStatus = [];
```
Comment (sanmai) [inline src/Metrics/MetricsCalculator.php:53]: Contrary to the above, the array-key type is significant here. We'll be using it and comparing against strings.

File: tests/phpunit/Logger/JsonLoggerTest.php (hunk 4 above)
Comment (sanmai) [reply tests/phpunit/Logger/JsonLoggerTest.php]: I got it back.

File: src/Event/Subscriber/MutationTestingResultsCollectorSubscriber.php (hunk 2 above)
Comment (sanmai) [reply src/Event/Subscriber/MutationTestingResultsCollectorSubscriber.php:46]: Again, key-type is insignificant here.

Comment (maks-rafalko) [reply src/Event/Subscriber/MutationTestingResultsCollectorSubscriber.php:46]: but what kind of integer keys are there? If they are natural numbers, we can use `list<Collector>`, if they are integers with "wholes", it would be better to use `array<int, Collector>`

IMO, `int` is better than implicit `array-key`

File: tests/phpunit/Logger/JsonLoggerTest.php [hunk 6]
```diff
@@ -246,12 +252,13 @@ public function metricsProvider(): iterable
 
         yield 'Non UTF-8 characters' => [
             false,
-            $this->createNonUtf8CharactersCalculator(),
+            new MetricsCalculator(2),
+            $this->createNonUtf8CharactersCollector(),
             [
                 'stats' => [
-                    'totalMutantsCount' => 1,
```
Comment (maks-rafalko) [inline tests/phpunit/Logger/JsonLoggerTest.php:252]: same question, why was it changed? (I mean probably it's ok, I just didn't get the change)

**Review by maks-rafalko — APPROVED** [review]

File: src/Event/Subscriber/MutationTestingResultsCollectorSubscriber.php (hunk 2 above)
Comment (sanmai) [reply src/Event/Subscriber/MutationTestingResultsCollectorSubscriber.php:46]: What I mean is that the API of `MutationTestingResultsCollectorSubscriber` does not care about keys.

`iterable<Collector>` could work.

File: tests/phpunit/Logger/JsonLoggerTest.php (hunk 6 above)
Comment (sanmai) [reply tests/phpunit/Logger/JsonLoggerTest.php:252]: It tests against empty `MetricsCalculator`, and it does that because it's a test of an edge-case of `JsonLogger`. We do not actually care about these numbers here.

**Comment by sanmai** [comment]
Thank you @maks-rafalko
